### PR TITLE
feat(apache): enhance TLS settings

### PR DESCRIPTION
**Pull request by fzipi**
Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

In the spirit of #103, enhance the TLS settings for Apache too.

Will check variable `APACHE_ALWAYS_TLS_REDIRECT` and if set to `on` will perform automatic redirection.

### CODE REVIEW

**Comment by fzipi** [comment]
Needed to do two different rewrite based on if there was a port or not. My tests for:
- variable `off`, PORT = 80
- variable `on`, PORT = 80
- variable `off`, PORT = 8080
- variable `on`, PORT = 8080

worked. All those are simple changes in the `apache-waf` section of the `docker-compose.yml` file.

**Comment by dune73** [comment]
I did not run this, but what I see looks good. Greatly expanding the SSL usability of the off-the-shelf docker container.

File: v2-apache/Dockerfile [hunk 1]
```diff
@@ -108,8 +108,16 @@ ENV ACCESSLOG=/var/log/apache2/access.log \
     REQ_HEADER_FORWARDED_PROTO='https' \
     SERVER_ADMIN=root@localhost \
     SERVER_NAME=localhost \
+    APACHE_ALWAYS_TLS_REDIRECT=off \
     SSL_ENGINE=on \
     SSL_PORT=443 \
+    SSL_PROTOCOL="all -SSLv3 -TLSv1 -TLSv1.1" \
+    SSL_PROXY_PROTOCOL="all -SSLv3 -TLSv1 -TLSv1.1" \
```
Comment (theseion) [inline v2-apache/Dockerfile:115]: Group proxy variables together.

File: v2-apache/Dockerfile-alpine [hunk 2]
```diff
@@ -123,8 +123,16 @@ ENV ACCESSLOG=/var/log/apache2/access.log \
     REQ_HEADER_FORWARDED_PROTO='https' \
     SERVER_ADMIN=root@localhost \
     SERVER_NAME=localhost \
+    APACHE_ALWAYS_TLS_REDIRECT=off \
     SSL_ENGINE=on \
     SSL_PORT=443 \
+    SSL_PROTOCOL="all -SSLv3 -TLSv1 -TLSv1.1" \
+    SSL_PROXY_PROTOCOL="all -SSLv3 -TLSv1 -TLSv1.1" \
```
Comment (theseion) [inline v2-apache/Dockerfile-alpine:130]: Group proxy variables together.

File: v2-apache/conf/extra/httpd-vhosts.conf [hunk 3]
```diff
@@ -29,6 +29,18 @@ SSLProxyCACertificateFile ${PROXY_SSL_CA_CERT}
 
 UseCanonicalName on
 
+<VirtualHost *:${PORT}>
+    RewriteEngine On
+    RewriteCond %{REQUEST_URI} !^/\.well\-known/acme\-challenge/
+    RewriteCond %{ENV:APACHE_ALWAYS_TLS_REDIRECT} on
+    RewriteCond %{HTTP_HOST} ^(.*):${PORT}$ [NC]
```
Comment (theseion) [inline v2-apache/conf/extra/httpd-vhosts.conf:36]: Why not use `SERVER_PORT` here?

File: v2-apache/conf/extra/httpd-vhosts.conf [hunk 4]
```diff
@@ -29,6 +29,18 @@ SSLProxyCACertificateFile ${PROXY_SSL_CA_CERT}
...
+    RewriteRule ^(.*)$ https://%1:${SSL_PORT}$1 [L,R=301]
+
+    RewriteCond %{REQUEST_URI} !^/\.well\-known/acme\-challenge/
+    RewriteCond %{ENV:APACHE_ALWAYS_TLS_REDIRECT} on
+    RewriteRule ^(.*)$ https://%{HTTP_HOST}:${SSL_PORT}$1 [L,R=301]
```
Comment (theseion) [inline v2-apache/conf/extra/httpd-vhosts.conf:41]: Why do you have two different rewrites here? Both end up redirecting to the same `(host, port)` tuple, correct?

**Review by theseion — CHANGES_REQUESTED** [review]
Two general comments:
1. I think you should add two columns to the table for "allowed values" and "default value". That should make reading (and writing) the table easier.
2. AFAICT, there's no way to terminate TLS at the proxy. As soon as you enable TLS on the WAF, the expectation is that the connection to the backend will also be TLS encrypted. I think it's often a viable setup to terminate TLS at the proxy and use plain `HTTP` to connect to the backend.

File: v2-apache/conf/extra/httpd-vhosts.conf (hunk 3 above)
Comment (fzipi) [reply v2-apache/conf/extra/httpd-vhosts.conf:36]: Is that a thing? Double checked docs and I can't find anything similar.

File: v2-apache/conf/extra/httpd-vhosts.conf (hunk 4 above)
Comment (fzipi) [reply v2-apache/conf/extra/httpd-vhosts.conf:41]: Yes. But it depends a lot if the PORT where the server is running is the default or a different one. That's why you need both rewrites.

File: README.md [hunk 5]
```diff
@@ -117,7 +121,16 @@ $ docker run -p 8080:80 -e SERVER_NAME=myhost my-modsec
 | REMOTEIP_INT_PROXY  | A string indicating the client intranet IP addresses trusted to present the RemoteIPHeader value (Default: `10.1.0.0/16`) | 
 | SERVER_ADMIN  | A string value indicating the address where problems with the server should be e-mailed (Default: `root@localhost`) | 
 | REQ_HEADER_FORWARDED_PROTO  | A string indicating the transfer protocol of the initial request (Default: `https`) | 
+| APACHE_ALWAYS_TLS_REDIRECT | A string value indicating if http should redirect to https (Allowed values: `on`, `off`. Default: `off`) | 
 | SSL_ENGINE  | A string indicating the SSL Engine Operation Switch (Default: `off`) | 
+| SSL_PORT  | Port number where the SSL enabled webserver is listening (Default: `443`) | 
+| SSL_PROTOCOL | A string for configuring the [usable SSL/TLS protocol versions](https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslprotocol). (Default `"all -SSLv3 -TLSv1 -TLSv1.1"`) |
```
[the remark this answers is not in the document]
Comment (fzipi) [reply README.md:127]: Done.

File: README.md [hunk 6]
```diff
@@ -117,7 +121,16 @@ $ docker run -p 8080:80 -e SERVER_NAME=myhost my-modsec
...
+| SSL_PROXY_PROTOCOL | A string for configuring the [proxy client SSL/TLS protocol versions](https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslproxyprotocol) (Default: `"all -SSLv3 -TLSv1 -TLSv1.1"`) |
```
[the remark this answers is not in the document]
Comment (fzipi) [reply README.md:128]: Grouped. Also sorted variables alphabetically.

File: README.md [hunk 7]
```diff
@@ -117,7 +121,16 @@ $ docker run -p 8080:80 -e SERVER_NAME=myhost my-modsec
...
+| SSL_SESSION_TICKETS | A string to enable or disable the use of [TLS session tickets](https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslsessiontickets) (RFC 5077). (Default: `off`) |
+| SSL_USE_STAPLING | A string indicating if [OSCP Stapling](https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslusestapling) should be used.(Default: `On`) |
```
[the remark this answers is not in the document]
Comment (fzipi) [reply README.md:130]: Done.

File: README.md [hunk 8]
```diff
@@ -117,7 +121,16 @@ $ docker run -p 8080:80 -e SERVER_NAME=myhost my-modsec
...
+| SSL_CIPHER_SUITE | A string indicating the cipher suite to use. Uses OpenSSL [list of cipher suites](https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set_ciphersuites.html) (Default: `"ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384"` |
+| SSL_PROXY_CIPHER_SUITE | A string indicating the proxy cipher suite to use (ciphers to use to connect to your backend). (Default `"ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384"` |
```
[the remark this answers is not in the document]
Comment (fzipi) [reply README.md:132]: Done.

File: README.md [hunk 9]
```diff
@@ -117,7 +121,16 @@ $ docker run -p 8080:80 -e SERVER_NAME=myhost my-modsec
...
+| SSL_HONOR_CIPHER_ORDER | A string indicating if the server should [honor the cipher list provided by the client](https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslhonorcipherorder). (Default: `off`) |
```
[the remark this answers is not in the document]
Comment (fzipi) [reply README.md:133]: Done.

File: docker-compose.yml [hunk 10]
```diff
@@ -8,14 +8,15 @@ services:
     environment:
       BACKEND: http://backend  # https://httpbin.org
       PORT: "8080"
-      SSL_PORT: "4443"
+      SSL_PORT: "8443"
       METRICS_ALLOW_FROM: All
+      APACHE_ALWAYS_TLS_REDIRECT: "on"
```
[the remark this answers is not in the document]
Comment (fzipi) [reply docker-compose.yml:13]: If you take a look at https://github.com/coreruleset/modsecurity-docker/blob/6c6450c45378b4c332561c86f4d33b75edd786ed/v2-apache/Dockerfile#L50, we are adding a generic certificate there.

File: v2-apache/Dockerfile (hunk 1 above)
Comment (fzipi) [reply v2-apache/Dockerfile:115]: Done.

File: v2-apache/Dockerfile-alpine (hunk 2 above)
Comment (fzipi) [reply v2-apache/Dockerfile-alpine:130]: Done

File: v2-apache/conf/extra/httpd-vhosts.conf (hunk 3 above)
Comment (theseion) [reply v2-apache/conf/extra/httpd-vhosts.conf:36]: See here https://httpd.apache.org/docs/2.4/mod/core.html#usecanonicalname and here https://httpd.apache.org/docs/current/expr.html.

File: v2-apache/conf/extra/httpd-vhosts.conf (hunk 4 above)
Comment (theseion) [reply v2-apache/conf/extra/httpd-vhosts.conf:41]: Ok. I don't understand it but I trust you know why :)

File: v2-apache/conf/extra/httpd-vhosts.conf (hunk 3 above)
Comment (fzipi) [reply v2-apache/conf/extra/httpd-vhosts.conf]: So the original problem was that HTTP_HOST comes from the client, and might include the PORT. That's why we needed two different. But forgot that we have `UseCanonicalNames` activated by default, so we can switch to SERVER_NAME and forget about HTTP_HOST.

Comment (theseion) [reply v2-apache/conf/extra/httpd-vhosts.conf]: I see. Now I understand why you had two different rewrite rules there.

**Review by theseion — COMMENTED** [review]
Looks good.

I still wonder whether there should be a setting for `SSLPROXY_ENGINE` (https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslproxyengine), to enable / disable TLS between proxy and backend. Right now, that connection will always be encrypted as soon as TLS redirect has been activated, but there's a good chance that the backend uses plain HTTP.

**Comment by fzipi** [comment]
Got it. We have it already, it is called `SSLProxyEngine ${PROXY_SSL}`. Maybe we can rename it, but could break for some people.

**Comment by theseion** [comment]
Sorry! The variable was in one of the collapsed parts, so I missed it... That's great then.

**Review by theseion — APPROVED** [review]

**Pull request merged**